LF: Update DAML-LF Governance process (#7926)
digital-asset/daml: daml-lf/governance.rst

==========================

These are the principles and rules by which changes to three data
formats are governed by the DAML Language team:

* language,
* values and,
* transaction nodes, and
* transactions.

Overview
--------

* The *DAML-LF language* is the typed core calculus that DAML gets
  compiled to, and that the ledger runs.

* *DAML-LF  serializable  values*  or   (DAML-LF  values  for  brevity
  (DAML-LF values for brevity in the restin the rest of this document)
  is  the  universe of  DAML-LF  values  that  can  be stored  on  the
  ledger.   Currently  nominal   raecords,  nominal   variants,  lists,
  optionals, maps, and primitive types.

* *DAML-LF transaction nodes* (DAML-LF nodes for brevity in the rest
   of this document) used to represent DAML actions described in the
   DAML ledger model. Currently there are *create*, *exercise*,
   *fetch*, and *lookup-by-key* nodes.
   
* *DAML-F transactions* are sequences of transaction nodes that
  defines an update to the ledger.
  
* The DAML-LF language is versioned using a major and minor component.
  Changes to the major component allow us to drop features, change the

>   semantics of existing features, or update the serialization format.

Author: I cannot really tell you, we have never created new major version. 
I just move this text around.

>   semantics of existing features, or update the serialization format.

Reviewer: I see. Then let's leave it as is for now.

  and operate on the same major version of the serialization format in
  a backward compatible way. For historical reason this scheme
  starts at `1.6`.

* DAML-LF values, nodes, and transactions are versioned using a common
  one component version scheme, called transaction version scheme.
  For historical reason this scheme starts at `10`.
  
* DAML-LF Language, values, nodes, and transactions are evolved
  together without ever breaking backwards compatibility.
  
* DAML-LF values, nodes, and transactions, and language are involved
  together, meaning the introduction of a new language version implies
  the introduction of a new transaction version, and reciprocally.

* A specification is provides for each major version language, and A
  common specification document are provided for values, nodes,
  transactions.  Moreover, a one-line summary of each version must be
  present in the ``daml-lf-X.proto`` and ``transaction.proto`` files.

* Every change to DAML-LF values, nodes, transactions or languages
  *must* be checked for governance rules conformance and approved must
  be approved by at least two persons among Rémy Haemmerle, Martin

>   Huschenbett and Sophia Mota.

Reviewer 2: ```suggestion
  Huschenbett and Sofia Faro.
```

  These checks are for governance and versioning only; the substance

>   of DAML-LF language changes should be made in consultation with the

Author: I dropped the paragraph.

  Language team.  Proposed changes can be filed as tickets under
  labels "component/daml-lf" and "discussion".

Language
--------

Some version bumps might not change the ``.proto`` at all, but just
add / change semantics in the specification. For example, the string

> format for decimal numbers in DAML-LF values might be made more

Reviewer: This sounds like we're breaking backward compatibility. Maybe we need to be more specific.

Author: Again it is just paragraph that I moved around. 
I am not sure how we can be more specific. 

``
For example, the string format for decimal numbers in DAML-LF values might be made more lenient in newer version  than in the original `1.10`.
``

Reviewer: Ok, if it was here before, then let's leave it as is.

lenient, resulting in a version bump with no ``.proto`` schema

> changes (but probably a comment).

Reviewer: ```suggestion
changes but only an additional comment.
```

Also note that the DAML-LF versioning is independent from the DAML
surface language versioning.

"dev" version
~~~~~~~~~~~~~

> Every DAML-LF major version includes a minor version, the *dev* version,

Reviewer: ```suggestion
Every DAML-LF major version includes a minor version called *dev*,
```

which we use as staging area for the next stable minor version of that
major version.

> The "dev" version can be changed freely without
> compatibility considerations to prior "dev" versions.  Since "dev" is

Reviewer: ```suggestion
The *dev* version can be changed freely without
compatibility considerations to prior *dev* versions.  Since *dev* is
```

always considered to be newer than every stable minor version of that
major version, it must be backward-compatible with all such stable
versions.

All newly-implemented minor version features or changes must be staged

> in the *dev* version for later release as a stable group of

Reviewer: ```suggestion
in the *dev* version for later release in a new stable minor version, which adds only
```

backward-compatible features.

The DAML-LF dev version is enabled in the sandbox and ledger server, but

> will never be emitted by damlc unless explicitly requested via

Reviewer: ```suggestion
will never be emitted by ``damlc`` unless explicitly requested via
```

``--target 1.dev`` or similar.

Associated Transaction version
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Except for `1.6`, `1.7`, and `1.8` each language version is associated

> to an unique transaction version, called the associated transaction

Reviewer: Aren't each of them assigned to a unique version as well, although that version is the same for all of them? The non-uniqueness only exists the other way around, in the sense that this version is not associated to a unique language version.

Author: fix.

version. Each time a new language version is introduced, it should be associated to
a new transaction version.

This mapping is clearly indicated in the DAML-LF specification
under the "Associated transaction versions" section.  Language
versions `1.6`, `1.7` and `1.8` are all associated to transaction
version `10`. Starting for language version `1.11`, each language version `1.x` is associated with transaction

> `1.11`

Reviewer: Shouldn't this be `1.x`?

Author: good catch.
